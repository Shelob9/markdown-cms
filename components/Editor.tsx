import Link from 'next/link';
import React,{ FC, forwardRef,  useRef, useState } from 'react';
import CmsLayout from './CmsLayout';
import { TextInput, TextInputProps } from './FieldAreas';
import MarkdownEditor from './MarkdownEditor';



const TitleField: FC<TextInputProps> = forwardRef((props, ref) => (
    <TextInput {...{
        ...props,
        required: true,
    }}
        //@ts-ignore
        ref={ref}
    />
));


export type saveData = { content: string, filePath: string };
export type editorProps = {
    onSave: (update: saveData) => Promise<any>;
    initialPath?: string;
    initialContent?: string;
    name?: string;
};

const Editor : FC<editorProps> = ({onSave,initialPath,initialContent,name}) => {
    let filePathRef = useRef<HTMLInputElement>();
    const [content, setContent] = useState(initialContent);
    const [isSaving, setISaving] = useState(false);
    const saveHandler = () => {
        setISaving(true);
        onSave({
            content,
            filePath:filePathRef.current.value
        }).then(() => {
            setISaving(false);
        })
    }
    return (
        <CmsLayout TopBar={() => (
            <>
                <button title={'Save'} onClick={saveHandler} className={'btn-primary'}>
                    {!isSaving ? 'Save' : <span className={'animate-spin h-5 w-5 mr-3'}>Saving</span>}
                </button>
                {name && <Link href={`/${name}`}>
                        View
                </Link>
                }
            </>
        )}>
            <TextInput
                label={'File Name'}
                ref={filePathRef}
                id={'file-name'}
                defaultValue={initialPath}
                disabled={isSaving}
            />
            <MarkdownEditor
                value={content}
                setValue={isSaving ? () => { } : setContent}
            /> 
        </CmsLayout>
    )
}

export default Editor;
  
