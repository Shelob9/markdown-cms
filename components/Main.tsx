import * as React from 'react';

const Main = ({ children }) => {
  return (
    <main className="container mx-auto flex-auto relative border-t border-gray-200 dark:border-gray-800">
      { children }
    </main>
  )
}

export default Main;
