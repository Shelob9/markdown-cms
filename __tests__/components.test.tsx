import * as React from 'react';

import Main from '../components/Main';

import { render,fireEvent, act } from '@testing-library/react';
import Editor from '../components/Editor';
test('Main renders', () => {
    const { container, getByText } = render(<Main><span>Hi Roy</span></Main>);
    expect(getByText('Hi Roy')).toBeTruthy();
    expect(container).toMatchSnapshot();
})

let eventFactory = (value: string) => {
    return {
        target: { value }
    };
}
describe('Editor', () => {

    it('calls onSave', () => {
        let onSave = jest.fn();
        const { getByLabelText,getByTitle } = render(<Editor onSave={onSave} />)
    
        act(() => {
            fireEvent.change(
                getByLabelText('Title'), eventFactory('title')
            );
        });

        act(() => {
            fireEvent.change(
                getByLabelText('Content'), eventFactory('content')
            );
        });

        act(() => {
            fireEvent.click(
                getByTitle('Save')
            );
        });
            
        expect(onSave).toBeCalledTimes(1);
        expect(onSave).toBeCalledWith('title', 'content');

            
    });
});