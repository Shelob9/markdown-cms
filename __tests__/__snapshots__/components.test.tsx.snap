// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main renders 1`] = `
<div>
  <main
    class="container mx-auto flex-auto relative border-t border-gray-200 dark:border-gray-800"
  >
    <span>
      Hi Roy
    </span>
  </main>
</div>
`;
